File: app/routers/sessions.py
from datetime import datetime, timezone
from fastapi import APIRouter, Depends
import aiosqlite

from app.database import get_db
from app.models.sessions import CorrelateRequest, CorrelateResponse
from app.services.slack import notify_session_correlated

router = APIRouter()


@router.post("/sessions/correlate", response_model=CorrelateResponse)
async def correlate_session(
    req: CorrelateRequest, db: aiosqlite.Connection = Depends(get_db)
):
    row = await db.execute_fetchall(
        "SELECT web_user_id FROM sessions WHERE id = ?", (req.session_id,)
    )
    if not row:
        return CorrelateResponse(success=False, correlated=False)

    web_user_id = row[0][0]
    now = datetime.now(timezone.utc).isoformat()

    await db.execute(
        "UPDATE sessions SET cli_user_id = ?, correlated_at = ? WHERE id = ?",
        (req.cli_user_id, now, req.session_id),
    )
    await db.commit()

    if web_user_id:
        await notify_session_correlated(req.session_id, web_user_id, req.cli_user_id)

    return CorrelateResponse(
        success=True, correlated=True, web_user_id=web_user_id
    )

File: src/handlers/sessions.rs
use axum::extract::State;
use axum::Json;
use chrono::Utc;
use sqlx::SqlitePool;

use crate::models::{CorrelateRequest, CorrelateResponse};
use crate::slack;

pub async fn correlate_session(
    State(pool): State<SqlitePool>,
    Json(req): Json<CorrelateRequest>,
) -> Json<CorrelateResponse> {
    let row: Option<(Option<String>,)> =
        sqlx::query_as("SELECT web_user_id FROM sessions WHERE id = ?")
            .bind(&req.session_id)
            .fetch_optional(&pool)
            .await
            .ok()
            .flatten();

    let (success, correlated, web_user_id) = match row {
        None => (false, false, None),
        Some((db_web_user_id,)) => {
            let now = Utc::now().to_rfc3339();
            sqlx::query("UPDATE sessions SET cli_user_id = ?, correlated_at = ? WHERE id = ?")
                .bind(&req.cli_user_id)
                .bind(&now)
                .bind(&req.session_id)
                .execute(&pool)
                .await
                .ok();

            if db_web_user_id.as_ref().map_or(false, |s| !s.is_empty()) {
                let session_id = req.session_id.clone();
                let wid = db_web_user_id.clone().unwrap();
                let cli_user_id = req.cli_user_id.clone();
                tokio::spawn(async move {
                    slack::notify_session_correlated(&session_id, &wid, &cli_user_id).await;
                });
            }

            (true, true, db_web_user_id)
        }
    };

    Json(CorrelateResponse {
        success,
        correlated,
        web_user_id,
    })
}
